packages/@react-aria/i18n/src/context.tsx

```diff
@@ -23,36 +23,65 @@ export interface I18nProviderProps {
 
 const I18nContext = React.createContext<Locale | null>(null);
 
+interface I18nProviderWithLocaleProps extends I18nProviderProps {
+  locale: string
+}
+
 /**
- * Provides the locale for the application to all child components.
+ * Internal component that handles the case when locale is provided.
  */
-export function I18nProvider(props: I18nProviderProps): JSX.Element {
-  let {locale, children} = props;
-  let defaultLocale = useDefaultLocale();
+function I18nProviderWithLocale(props: I18nProviderWithLocaleProps): JSX.Element {
+  let {locale, children} = props;  
+  let value: Locale = React.useMemo(() => ({
+    locale,
+    direction: isRTL(locale) ? 'rtl' : 'ltr'
+  }), [locale]);
+
+  return (
+    <I18nContext.Provider value={value}>
+      {children}
+    </I18nContext.Provider>
+  );
+}
 
-  let value: Locale = React.useMemo(() => {
-    if (!locale) {
-      return defaultLocale;
-    }
+interface I18nProviderWithDefaultLocaleProps {
+  children: ReactNode
+}
 
-    return {
-      locale,
-      direction: isRTL(locale) ? 'rtl' : 'ltr'
-    };
-  }, [defaultLocale, locale]);
+/**
+ * Internal component that handles the case when no locale is provided.
+ */
+function I18nProviderWithDefaultLocale(props: I18nProviderWithDefaultLocaleProps): JSX.Element {
+  let {children} = props;
+  let defaultLocale = useDefaultLocale();
 
   return (
-    <I18nContext.Provider value={value}>
+    <I18nContext.Provider value={defaultLocale}>
       {children}
     </I18nContext.Provider>
   );
 }
 
+/**
+ * Provides the locale for the application to all child components.
+ */
+export function I18nProvider(props: I18nProviderProps): JSX.Element {
+  let {locale, children} = props;
+
+  // Conditionally render different components to avoid calling useDefaultLocale.
+  // This is necessary because useDefaultLocale triggers a re-render.
+  if (locale) {
+    return <I18nProviderWithLocale locale={locale} children={children} />;
+  }
+
+  return <I18nProviderWithDefaultLocale children={children} />;
+}
+
 /**
  * Returns the current locale and layout direction.
  */
 export function useLocale(): Locale {
-  let defaultLocale = useDefaultLocale();
+  let defaultLocale = getDefaultLocale();
```

Comment:
I think this one has to stay so the locale doesn't become stale?

Maybe it'd be a good idea to add some tests making use of `window.addEventListener('languagechange'` it doesn't look like we have any right now.